### PR TITLE
fix: enable personal account.

### LINKED ISSUE #5393: Impossible to sign in to personal account to create a publish profile
## Describe the bug

To publish the bot, I clicked on "Add new publish profile" in the configuration section. I selected the option to create new Azure resources which requires signing in.

First bug is that the input doesn't take "@" as input for some reason, so I had to copy/paste my email address to enter it. I'm using my personal email address (@hotmail.fr)
2nd and most important problem: I have a message saying "We couldn't find a work or school account with that email address." 
I do have an Azure Subscription linked to that email address so I don't understand what's going on.

## Version

v1.3.0
image

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## To Reproduce

1. Create a bot on Composer (I used the LUIS + QnA  template)
2. Try creating a new publish profile
3. Try signing in with a personal account

## Expected behavior

I would expect to be able to sign in with my personal account linked to an Azure subscription.

## Screenshot

image

### LINKED ISSUE #5706: Unable to login with personal account
Hi Team,

I am not able to login with my personal account for creating a publish profile on composer v1.3. Any help re this would be much appreciated.
image

Cheers.

### PR DESCRIPTION
**Pull request by VanyLaw**
## Description

fix login fail if user used personal account.

## Task Item

close #5706, close #5393

### CODE REVIEW

File: extensions/azurePublish/src/components/util.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      result.push(AzureResourceTypes.COSMOSDB);
+    }
+    if(config.settings?.blobStorage?.connectionString){
+      result.push(AzureResourceTypes.BLOBSTORAGE);
+    }
+    return result;
+  } else return [];
+}
+
+export const iconStyle = (required) => {
```
Comment (hatpick) [inline extensions/azurePublish/src/components/util.tsx:63]: Please move this to wherever it's used, I don't see it being used in any other place, so no need to be exported.

(change name to .ts after you moved it)

File: extensions/azurePublish/src/components/util.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import { getTenants, getARMTokenForTenant } from '@bfc/extension-client';
+import jwtDecode from 'jwt-decode';
+import { SharedColors } from '@uifabric/fluent-theme';
+import { AzureResourceTypes } from '../types';
+
+export function decodeToken(token: string) {
```
Comment (hatpick) [inline extensions/azurePublish/src/components/util.tsx:6]: nit: is there any particular reason that some of the functions in this file are arrow and some are using traditional function definition? (prefer arrow)

File: extensions/azurePublish/src/components/util.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import { getTenants, getARMTokenForTenant } from '@bfc/extension-client';
```
Comment (hatpick) [inline extensions/azurePublish/src/components/util.tsx:1]: this file is not formatted correctly. e.g., the copyright missing on top, spacings are off, etc.

File: Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx [hunk 4]
```diff
@@ -140,8 +140,17 @@ export const PublishProfileDialog: React.FC<PublishProfileDialogProps> = (props)
...
+          let tenantId = getTenantIdFromCache();
+          if (!tenantId) {
+            const tenants = await AuthClient.getTenants();
+            tenantId = tenants[0].tenantId;
+            setTenantId(tenantId);
+          }
+          console.log('tenant-------', tenantId);
+          arm = await AuthClient.getARMTokenForTenant(tenantId);
           graph = await AuthClient.getAccessToken(graphScopes);
+          console.log('arm---------', arm);
```
Comment (hatpick) [inline Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx:152]: remove logs?

File: Composer/packages/client/src/pages/botProject/create-publish-profile/ProfileFormDialog.tsx [hunk 5]
```diff
@@ -135,7 +135,13 @@ export const ProfileFormDialog: React.FC<ProfileFormDialogProps> = (props) => {
       let arm, graph;
       if (!isGetTokenFromUser()) {
         // login or get token implicit
-        arm = await AuthClient.getAccessToken(armScopes);
+        let tenantId = getTenantIdFromCache();
+        if (!tenantId) {
+          const tenants = await AuthClient.getTenants();
+          tenantId = tenants[0].tenantId;
```
Comment (tonyanziano) [inline Composer/packages/client/src/pages/botProject/create-publish-profile/ProfileFormDialog.tsx:141]: what if `tenants` is an empty array

File: Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx [hunk 6]
```diff
@@ -140,8 +140,17 @@ export const PublishProfileDialog: React.FC<PublishProfileDialogProps> = (props)
         let arm, graph;
         if (!isGetTokenFromUser()) {
           // login or get token implicit
-          arm = await AuthClient.getAccessToken(armScopes);
+          let tenantId = getTenantIdFromCache();
+          if (!tenantId) {
+            const tenants = await AuthClient.getTenants();
+            tenantId = tenants[0].tenantId;
```
Comment (tonyanziano) [inline Composer/packages/client/src/pages/botProject/create-publish-profile/PublishProfileDialog.tsx:146]: what if `tenants` is an empty array?

File: Composer/packages/client/src/pages/publish/Publish.tsx [hunk 7]
```diff
@@ -219,7 +224,13 @@ const Publish: React.FC<RouteComponentProps<{ projectId: string; targetName?: st
     if (isGetTokenFromUser()) {
       token = getTokenFromCache('accessToken');
     } else {
-      token = await AuthClient.getAccessToken(armScopes);
+      let tenant = getTenantIdFromCache();
+      if (!tenant) {
+        const tenants = await AuthClient.getTenants();
+        tenant = tenants[0].tenantId;
```
Comment (tonyanziano) [inline Composer/packages/client/src/pages/publish/Publish.tsx:230]: what if `tenants` is an empty array?

File: Composer/packages/electron-server/src/auth/oneAuthService.ts [hunk 8]
```diff
@@ -181,6 +185,12 @@ export class OneAuthInstance extends OneAuthBase {
   }
 
   public async getARMTokenForTenant(tenantId: string): Promise<string> {
+    // sign in arm account.
+    if (!this.signedInARMAccount) {
+      const signInParams = new this.oneAuth.AuthParameters(DEFAULT_AUTH_SCHEME, ARM_AUTHORITY, ARM_RESOURCE, '', '');
+      const result: OneAuth.AuthResult = await this.oneAuth.signInInteractively('', signInParams, '');
+      this.signedInARMAccount = result.account;
+    }
```
Comment (tonyanziano) [inline Composer/packages/electron-server/src/auth/oneAuthService.ts:193]: update tests to cover this

File: Composer/packages/electron-server/src/auth/oneAuthService.ts [hunk 9]
```diff
@@ -105,6 +105,10 @@ export class OneAuthInstance extends OneAuthBase {
         log('Mac development env detected. Getting access token using interactive sign in instead of silently.');
         return this.TEMPORARY_getAccessTokenOnMacDev(params);
       }
+      // if Account not exist, use arm account. in case the window popup again
+      if (!this.signedInAccount) {
+        this.signedInAccount = this.signedInARMAccount;
+      }
```
Comment (tonyanziano) [inline Composer/packages/electron-server/src/auth/oneAuthService.ts:111]: cover this in a test case

**Review by tonyanziano — CHANGES_REQUESTED** [review]

**Comment by luhan2017** [comment]
Show an error instead of spinning forever if no subscription under this account/tenant. 

image

**Review by tonyanziano — APPROVED** [review]
👍

**Review by luhan2017 — APPROVED** [review]

**Pull request merged**